## Part 1: Preprocessing

In [134]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [135]:
# Determine the number of unique values in each column
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [155]:
# Let's look at the unique value counts for attrition and departments
attrition_df['Attrition'].value_counts(), attrition_df['Department'].value_counts()

(Attrition
 No     1233
 Yes     237
 Name: count, dtype: int64,
 Department
 Research & Development    961
 Sales                     446
 Human Resources            63
 Name: count, dtype: int64)

In [136]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]


In [137]:
# Create a list of at least 10 column names to use as X data
x_columns = ['Age', 'DistanceFromHome', 'Education', 'JobSatisfaction', 'OverTime', 'StockOptionLevel',
    'WorkLifeBalance', 'YearsAtCompany', 'YearsSinceLastPromotion', 'NumCompaniesWorked']

# Create X_df using your selected columns
X_df = attrition_df[x_columns]


# Show the data types for X_df
X_df.dtypes



Age                         int64
DistanceFromHome            int64
Education                   int64
JobSatisfaction             int64
OverTime                   object
StockOptionLevel            int64
WorkLifeBalance             int64
YearsAtCompany              int64
YearsSinceLastPromotion     int64
NumCompaniesWorked          int64
dtype: object

In [138]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=42)


In [139]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary
X_train['OverTime'] = X_train['OverTime'].map({'No': 0, 'Yes': 1}).astype('int64')
X_test['OverTime'] = X_test['OverTime'].map({'No': 0, 'Yes': 1}).astype('int64')


# Check the data types for X_train and X_test
print("X_train dtypes:")
print(X_train.dtypes)
print("\nX_test dtypes:")
print(X_test.dtypes)



X_train dtypes:
Age                        int64
DistanceFromHome           int64
Education                  int64
JobSatisfaction            int64
OverTime                   int64
StockOptionLevel           int64
WorkLifeBalance            int64
YearsAtCompany             int64
YearsSinceLastPromotion    int64
NumCompaniesWorked         int64
dtype: object

X_test dtypes:
Age                        int64
DistanceFromHome           int64
Education                  int64
JobSatisfaction            int64
OverTime                   int64
StockOptionLevel           int64
WorkLifeBalance            int64
YearsAtCompany             int64
YearsSinceLastPromotion    int64
NumCompaniesWorked         int64
dtype: object


In [140]:
# Create a StandardScaler
scaler = StandardScaler()


# Fit the StandardScaler to the training data
scaler.fit(X_train)


# Scale the training and testing data
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [141]:
from sklearn.preprocessing import OneHotEncoder

# Create a OneHotEncoder for the Department column
encoder_dept = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
encoder_dept.fit(y_train[['Department']])


# Create two new variables by applying the encoder
# to the training and testing data
y_train_encoded_dept = encoder_dept.transform(y_train[['Department']])
y_test_encoded_dept = encoder_dept.transform(y_test[['Department']])

print(y_train_encoded_dept)


[[0. 1. 0.]
 [0. 0. 1.]
 [0. 0. 1.]
 ...
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [142]:
# Create a OneHotEncoder for the Attrition column
encoder_attr = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
encoder_attr.fit(y_train[['Attrition']])

# Create two new variables by applying the encoder
# to the training and testing data
y_train_encoded_attr = encoder_attr.transform(y_train[['Attrition']])
y_test_encoded_attr = encoder_attr.transform(y_test[['Attrition']])

print(y_train_encoded_attr)


[[1. 0.]
 [1. 0.]
 [1. 0.]
 ...
 [0. 1.]
 [1. 0.]
 [1. 0.]]


## Part 2: Create, Compile, and Train the Model

In [143]:
# Find the number of columns in the X training data.
num_cols = X_train_scaled.shape[1]
print(f"Number of columns in X training data: {num_cols}")


# Create the input layer
input = layers.Input(shape=(num_cols,), name='input')


# Create at least two shared layers
shared_1 = layers.Dense(64, activation='relu', name='shared_1')(input)
shared_2 = layers.Dense(128, activation='relu', name='shared_2')(shared_1)


Number of columns in X training data: 10


In [144]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
dept_hidden = layers.Dense(32, activation='relu', name='department_hidden')(shared_2)

# Create the output layer
dept_output = layers.Dense(y_train_encoded_dept.shape[1], activation='softmax', name='department_output')(dept_hidden)


In [145]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attr_hidden = layers.Dense(32, activation='relu', name='attrition_hidden')(shared_2)


# Create the output layer
attr_output = layers.Dense(y_train_encoded_attr.shape[1], activation='sigmoid', name='attrition_output')(attr_hidden)


In [146]:
# Create the model
model = Model(inputs=input, outputs=[dept_output, attr_output])


# Compile the model
model.compile(optimizer='adam',
              loss={'department_output': 'categorical_crossentropy', 'attrition_output': 'binary_crossentropy'},
              metrics={'department_output': 'accuracy', 'attrition_output': 'accuracy'})


# Summarize the model
model.summary()


Model: "functional_9"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input (InputLayer)  │ (None, 10)        │          0 │ -                 │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ shared_1 (Dense)    │ (None, 64)        │        704 │ input[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ shared_2 (Dense)    │ (None, 128)       │      8,320 │ shared_1[0][0]    │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ department_hidden   │ (None, 32)        │      4,128 │ shared_2[0][0]    │
│ (Dense)             │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attrition_hidden    │ (None, 32)        │      4,128 │ shared_2[0][0]    │
│ (Dense)             │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ department_output   │ (None, 3)         │         99 │ department_hidde… │
│ (Dense)             │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attrition_output    │ (None, 2)         │         66 │ attrition_hidden… │
│ (Dense)             │                   │            │                   │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 17,445 (68.14 KB)

 Trainable params: 17,445 (68.14 KB)

 Non-trainable params: 0 (0.00 B)

In [147]:
# Train the model
history = model.fit(
    X_train_scaled, 
    [y_train_encoded_dept, y_train_encoded_attr],
    epochs=100,
    batch_size=25,
    validation_split=0.2)

                    # validation_data=(X_test_scaled, [y_test_encoded_dept, y_test_encoded_attr]))


Epoch 1/100
36/36 ━━━━━━━━━━━━━━━━━━━━ 4s 18ms/step - attrition_output_accuracy: 0.6186 - attrition_output_loss: 0.6408 - department_output_accuracy: 0.6301 - department_output_loss: 0.8807 - loss: 1.5216 - val_attrition_output_accuracy: 0.7873 - val_attrition_output_loss: 0.5366 - val_department_output_accuracy: 0.6063 - val_department_output_loss: 0.8677 - val_loss: 1.4079
Epoch 2/100
36/36 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - attrition_output_accuracy: 0.8431 - attrition_output_loss: 0.4497 - department_output_accuracy: 0.6587 - department_output_loss: 0.7555 - loss: 1.2057 - val_attrition_output_accuracy: 0.7873 - val_attrition_output_loss: 0.4867 - val_department_output_accuracy: 0.6063 - val_department_output_loss: 0.8768 - val_loss: 1.3681
Epoch 3/100
36/36 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - attrition_output_accuracy: 0.8204 - attrition_output_loss: 0.4348 - department_output_accuracy: 0.6428 - department_output_loss: 0.7545 - loss: 1.1898 - val_attrition_output_accuracy: 0.7873 -

In [149]:
# Evaluate the model with the testing data
test_results = model.evaluate(X_test_scaled, {'department_output': y_test_encoded_dept, 'attrition_output': y_test_encoded_attr})
test_results


12/12 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - attrition_output_accuracy: 0.8223 - attrition_output_loss: 1.4441 - department_output_accuracy: 0.5852 - department_output_loss: 3.3580 - loss: 4.8052  


[4.29840087890625,
 3.116145133972168,
 1.1616426706314087,
 0.83152174949646,
 0.5978260636329651]

In [153]:
# Print the accuracy for both department and attrition
print(f"Attrition predictions accuracy: {test_results[3]}")
print(f"Department predictions accuracy: {test_results[4]}")

Attrition predictions accuracy: 0.83152174949646
Department predictions accuracy: 0.5978260636329651


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. The dataset involves predicting attrition and departments, which are both imbalanced. In such cases, accuracy can be misleading, because a model could predict the majority class and still appear accurate. 
- Attrition
  - No     1233
   - Yes     237
- Department
  - Research & Development    961
  - Sales                     446
  - Human Resources            63

2. I used softmax for Department because it returns a probability distribution across multiple categories. I also used sigmoid for Attrition since it is  ideal for binary Attrition

3. To improve this model, I could
- Evaluate class imbalance for Attrition and Department by applying techniques like SMOTE (oversampling), Class weighting, or Undersampling.
- Tune model architecture by adding more hidden layers or neurons or trying Dropout layers to reduce overfitting.
- Hyperparameter tuning like number of epochs and batch sizes.
- Use other metrics like F1-score, Confusion matrix, or even AUC-ROC for binary classification.
- Try encoding other categorical variables.